File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <limits.h>

#define LL long long

using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = LLONG_MAX / 2;
    LL left = 0, right = 100000001L * n;
    while (left <= right) {
        LL mid = (left + right) / 2;
        LL passedPerson = 0;
        for (int t : times) {
            passedPerson += mid / t;
        }
        if (passedPerson >= n) {
            right = mid - 1;
            answer = min(answer, mid);
        } else {
            left = mid + 1;
        }
    }
    return answer;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
class Solution {
    
    public long solution(int n, int[] times) {
        long answer = 0;
        long min = 1000000001;
        for (int i = 0; i < times.length; i++) {
            min = Math.min(min, times[i]);
        }
        long left = 0, right = min * n;
        while (left < right) {
            long mid = left + (right - left) / 2;
            
            if (checkPossibility(n, times, mid)) {
                answer = mid;
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        
        return answer;
    }
    
    static boolean checkPossibility(int n, int[] times, long t) {
        long passed = 0;
        for (int i = 0; i < times.length; i++) {
            passed += t / times[i];
        }
        
        return passed >= n;
    }
}